extensions/mssql/src/credentialstore/sqlCredentialService.ts

```diff
@@ -0,0 +1,144 @@
+			private _secretStorage: SecretStorage;
+			private _useNativeCredentialService: boolean;
+			private _passwordsMigrated: boolean;
+
+			constructor(client: SqlOpsDataClient) {
+				super(context, client);
+				this._secretStorage = context.extensionContext.secrets;
+				this._passwordsMigrated = context.extensionContext.globalState.get(Utils.configPasswordsMigrated);
+				this._useNativeCredentialService = Utils.useNativeCredentialsEnabled();
+			}
+
+			override fillClientCapabilities(capabilities: ClientCapabilities): void {
+				Utils.ensure(Utils.ensure(capabilities, 'credentials')!, 'credentials')!.dynamicRegistration = true;
+			}
+
+			override async initialize(capabilities: ServerCapabilities): Promise<void> {
+				this.register(this.messages, {
+					id: UUID.generateUuid(),
+					registerOptions: undefined
+				});
+				if (this._useNativeCredentialService && !this._passwordsMigrated) {
+					await this.migratePasswords();
+					this._passwordsMigrated = true;
+					await context.extensionContext.globalState.update(Utils.configPasswordsMigrated, this._passwordsMigrated);
+				}
+			}
+
+			/**
+			 * Removes a credential for a given connection profile from sqltoolsservice
+			 * and adds it to secret storage
+			 */
+			private async migrateCredential(conn: azdata.connection.ConnectionProfile): Promise<boolean> {
+				const credentialId = Utils.formatCredentialId(conn);
+				// read password from every saved password connection
+				const credential = await this._client.sendRequest(Contracts.ReadCredentialRequest.type, { credentialId, password: undefined });
+				if (credential.password) {
+					const password = credential.password;
+					// save it in secret store
+					await this._secretStorage.store(credentialId, password);
+					// check if it's saved
+					const savedPassword = await this._secretStorage.get(credentialId);
+					if (savedPassword === password) {
+						// delete from tools service
+						const result = await this._client.sendRequest(Contracts.DeleteCredentialRequest.type,
+							{ credentialId, password: undefined });
+						return result;
+					} else {
+						return false;
+					}
+				} else {
+					return false;
+				}
+			}
+
+			/**
+			 *
+			 * @returns Migrates all saved credentials to the native credential system
+			 */
+			private async migratePasswords(): Promise<void> {
+				const connections = await azdata.connection.getConnections(false);
+				const savedPasswordConnections = connections.filter(conn => conn.savePassword === true);
+				for (let i = 0; i < savedPasswordConnections.length; i++) {
+					let conn = savedPasswordConnections[i];
+					await this.migrateCredential(conn);
+				}
+				await Utils.removeCredentialFile();
+			}
+
+			protected override registerProvider(options: any): Disposable {
+				let readCredential = async (credentialId: string): Promise<azdata.Credential> => {
+					if (this._useNativeCredentialService && this._passwordsMigrated) {
```

Comment:
If useNativeCredentialService is true then I don't think we should ever be using the STS one. I would suggest having a deferred promise for the passwordsMigrated that you then await on so that it'll ensure that's all done before attempting to retrieve the credential.

Reply:
This is just for the first time when the credentials need to be migrated. It's a one time thing only since the passwordsMigrated will be set after the first time.

Reply:
Yes, but I'm saying that if the migration isn't completed by the time someone calls readCredential then we may end up in an invalid state where we go to STS to fetch a password but it doesn't exist there anymore (because it was in the process of being migrated).

Given that we now await the migration on startup (and block activation until that's done) is this check for the passwords being migrated even necessary?

Reply:
Like you said, we always migrate passwords at initialization, so that situation shouldn't occur.

We need the passwords migrated variable since it allows us to not initialize the STS credential store for the next time when a user opens ADS. It's only in the migration scenario we would need both secret stores initialized, otherwise we use either of the two.